File: frontend/src/services/pdfService.ts
import api from "./api";
import type { Resource } from "@/types";

export const pdfService = {
  async upload(
    file: File,
    data: {
      title: string;
      subject_id: string;
      chapter_id?: string;
      description?: string;
      tags?: string;
      importance?: string;
    }
  ) {
    const formData = new FormData();
    formData.append("file", file);
    formData.append("title", data.title);
    formData.append("subject_id", data.subject_id);
    if (data.chapter_id) formData.append("chapter_id", data.chapter_id);
    if (data.description) formData.append("description", data.description);
    if (data.tags) formData.append("tags", data.tags);
    if (data.importance) formData.append("importance", data.importance);

    const response = await api.post<Resource>("/api/pdfs/upload", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });
    return response.data;
  },

  async getSignedUrl(resourceId: string) {
    const response = await api.get<{ signed_url: string; resource_id: string }>(
      `/api/pdfs/${resourceId}/signed-url`
    );
    return response.data;
  },

  async delete(resourceId: string) {
    await api.delete(`/api/pdfs/${resourceId}`);
  },
};

File: backend/app/services/pdf_service.py
from sqlalchemy.orm import Session
from fastapi import UploadFile, HTTPException, status
from app.models.resource import Resource, ResourceTag
from app.constants.enums import ResourceType, ResourceStatus, Importance, IMPORTANCE_VALUES
from app.models.subject import Subject
from app.utils.supabase import upload_pdf_to_storage, delete_pdf_from_storage, get_signed_url


class PDFService:
    def __init__(self, db: Session):
        self.db = db

    def upload(
        self,
        user_id: str,
        file: UploadFile,
        title: str,
        subject_id: str,
        chapter_id: str = None,
        description: str = None,
        tags: list[str] = None,
        importance: str = Importance.NORMAL.value,
    ) -> Resource:
        if importance not in IMPORTANCE_VALUES:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail=f"Invalid importance '{importance}'. Must be one of: {', '.join(sorted(IMPORTANCE_VALUES))}",
            )
        subject = (
            self.db.query(Subject)
            .filter(Subject.id == subject_id, Subject.user_id == user_id)
            .first()
        )
        if not subject:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Subject not found",
            )

        pdf_url, file_name, file_size = upload_pdf_to_storage(user_id, file)

        resource = Resource(
            title=title,
            description=description,
            resource_type=ResourceType.PDF.value,
            status=ResourceStatus.NOT_STARTED,
            importance=importance,
            pdf_url=pdf_url,
            file_name=file_name,
            file_size=file_size,
            user_id=user_id,
            subject_id=subject_id,
            chapter_id=chapter_id,
        )

        self.db.add(resource)
        self.db.flush()

        if tags:
            for tag_name in tags:
                tag = ResourceTag(resource_id=resource.id, tag=tag_name.strip().lower())
                self.db.add(tag)

        self.db.commit()
        self.db.refresh(resource)
        return resource

    def get_signed_pdf_url(self, user_id: str, resource_id: str) -> str:
        resource = (
            self.db.query(Resource)
            .filter(
                Resource.id == resource_id,
                Resource.user_id == user_id,
                Resource.resource_type == ResourceType.PDF.value,
            )
            .first()
        )
        if not resource:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="PDF not found",
            )

        if not resource.file_name:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="PDF file not found in storage",
            )

        signed_url = get_signed_url(user_id, resource.file_name)
        if not signed_url:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to generate signed URL",
            )

        return signed_url

    def delete(self, user_id: str, resource_id: str) -> None:
        resource = (
            self.db.query(Resource)
            .filter(
                Resource.id == resource_id,
                Resource.user_id == user_id,
                Resource.resource_type == ResourceType.PDF.value,
            )
            .first()
        )
        if not resource:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="PDF not found",
            )

        if resource.file_name:
            delete_pdf_from_storage(user_id, resource.file_name)

        self.db.delete(resource)
        self.db.commit()
